xnmt/translator.py

```diff
@@ -191,3 +194,183 @@ def html_report(self, context=None):
     return html
 
 
+class TransformerTranslator(DefaultTranslator):
+  yaml_tag = u'!TransformerTranslator'
+
+  def __init__(self, src_embedder, encoder, attender, trg_embedder, decoder, input_dim=512):
+    '''Constructor.
+
+    :param src_embedder: A word embedder for the input language
+    :param encoder: An encoder to generate encoded inputs
+    :param attender: An attention module
+    :param trg_embedder: A word embedder for the output language
+    :param decoder: A decoder
+    '''
+    register_handler(self)
+    self.src_embedder = src_embedder
+    self.encoder = encoder
+    self.attender = attender
+    self.trg_embedder = trg_embedder
+    self.decoder = decoder
+    self.input_dim = input_dim
+    self.scale_emb = self.input_dim ** 0.5
+    self.initialize_position_encoding(500, input_dim)  # TODO: parametrize this
+
+  def initialize_generator(self, **kwargs):
+    if kwargs.get("len_norm_type", None) is None:
+      len_norm = xnmt.length_normalization.NoNormalization()
+    else:
+      len_norm = xnmt.serializer.YamlSerializer().initialize_object(kwargs["len_norm_type"])
+    search_args = {}
+    if kwargs.get("max_len", None) is not None:
+      search_args["max_len"] = kwargs["max_len"]
+      self.max_len = kwargs.get("max_len", 50)
+    if kwargs.get("beam", None) is None:
+      self.search_strategy = GreedySearch(**search_args)
+    else:
+      search_args["beam_size"] = kwargs.get("beam", 1)
+      search_args["len_norm"] = len_norm
+      # self.search_strategy = TransformerBeamSearch(**search_args)
+    self.report_path = kwargs.get("report_path", None)
+    self.report_type = kwargs.get("report_type", None)
+
+  def make_attention_mask(self, source_block, target_block):
+    mask = (target_block[:, None, :] <= 0) * \
+           (source_block[:, :, None] <= 0)
+    # (batch, source_length, target_length)
+    return mask
+
+  def make_history_mask(self, block):
+    batch, length = block.shape
+    arange = np.arange(length)
+    history_mask = (arange[None,] <= arange[:, None])[None,]
+    history_mask = np.broadcast_to(history_mask, (batch, length, length))
+    return history_mask
+
+  def mask_embeddings(self, embeddings, mask):
+    """
+    We convert the embeddings of masked input sequence to zero vector
+    """
+    (embed_dim, _), _ = embeddings.dim()
+    temp_mask = np.repeat(1. - mask[:, None, :], embed_dim, axis=1)
+    temp_mask = dy.inputTensor(np.moveaxis(temp_mask, [1, 0, 2], [0, 2, 1]), batched=True)
+    embeddings = dy.cmult(embeddings, temp_mask)
+    return embeddings
+
+  def initialize_position_encoding(self, length, n_units):
```

Comment:
This looks like reusable code and could optionally be pulled out into a PositionalEmbedder that takes a base_embedder and adds the positional encoding on top.

Reply:
Actually, I was just looking into this because I need this also, and making a PositionalEmbedder seems not as easy as I thought because of param sharing, so we should leave this as is for now.